amazeeio/lagoon
# Get CI running using helm charts

# Checklist

- [ ] Affected Issues have been mentioned in the Closing issues section
- [ ] Documentation has been written/updated
- [x] PR title is ready for changelog and subsystem label(s) applied

This PR refactors the Jenkins pipeline to get CI running using the Lagoon helm charts. The pipeline now uses [declarative syntax](https://www.jenkins.io/doc/book/pipeline/syntax/#declarative-pipeline) so it's a bit shorter and hopefully easier to understand. It also makes it easier to run a test [matrix](https://www.jenkins.io/doc/book/pipeline/syntax/#declarative-matrix) in future (e.g. multiple k8s versions).

Where the workflow has changed, it now does this:

* Build the images
* Push the images to `testlagoon/*`
* Create a kind cluster
* Copy the built images into the cluster cache (avoid pulling layers down again)
* Run tests in the cluster

The steps following a successful test run should be the same.

Because of the way that the [chart-testing](https://github.com/helm/chart-testing) tool works, we no longer have live log messages in Jenkins. Instead the test job will print a command which sets up a local `$KUBECONFIG` and a ssh port-forward so that you can interact with the `kind` cluster while it's running to tail logs, watch the tests run etc. Check the logs of the `Running tests on kind cluster` stage to see something like this:

![Screenshot from 2020-12-03 10-52-50](https://user-images.githubusercontent.com/861778/100958593-9bf6c600-3557-11eb-814e-35bb8fb0f4b3.png)

The things that I'd like special review attention on:

* Does the `Jenkinsfile` logic look correct compared to the old pipeline? Particularly around `SKIP_IMAGE_PUBLISH`, and when to push to `uselagoon/*`?
* Do the other workflows that use the `Makefile` still work? I don't do much with e.g. `docker-compose`, so could someone who uses that functionality confirm that it still works?
* Is there documentation that needs updating?
* Have I missed any test suites that need to be included and run?

# Closing issues

n/a
